File: cli/schedule.py
"""Interval checkpoint sampling — mirrors src/lib/schedule.ts."""

from __future__ import annotations

from datetime import datetime, timedelta
from typing import TypedDict

from .geo import Point, cumulative_distances_m, interpolate_along
from .routing import RouteResult


class Checkpoint(TypedDict, total=False):
    id: str
    label: str
    coord: Point
    eta: datetime
    offset_min: int
    cumulative_km: float
    kind: str  # 'origin' | 'interval' | 'destination'
    temp_c: float
    precip_mm: float
    icon: str
    hazard: bool


def sample_checkpoints(
    route: RouteResult,
    start: datetime,
    interval_min: int,
    origin_label: str,
    destination_label: str,
) -> list[Checkpoint]:
    total_min = route.duration_s / 60
    cum_m = cumulative_distances_m(route.polyline)
    total_km = route.distance_m / 1000

    offsets: list[float] = [0.0]
    t = float(interval_min)
    while t < total_min:
        offsets.append(t)
        t += interval_min
    offsets.append(total_min)

    checkpoints: list[Checkpoint] = []
    last = len(offsets) - 1
    for i, offset_min in enumerate(offsets):
        frac = offset_min / total_min if total_min > 0 else 0.0
        if i == 0:
            coord, kind, label = route.polyline[0], "origin", origin_label
        elif i == last:
            coord, kind, label = route.polyline[-1], "destination", destination_label
        else:
            coord = interpolate_along(route.polyline, cum_m, frac * route.distance_m)
            kind = "interval"
            label = f"Approx. location after {round(offset_min)} min"
        checkpoints.append(
            Checkpoint(
                id=f"cp-{i}",
                label=label,
                coord=coord,
                eta=start + timedelta(minutes=offset_min),
                offset_min=round(offset_min),
                cumulative_km=round(frac * total_km, 2),
                kind=kind,
                hazard=False,
            )
        )
    return checkpoints

File: src/lib/schedule.ts
import type { Checkpoint, RouteResult } from '../types'
import { cumulativeDistancesM, interpolateAlong } from './geo'

export interface SampleOptions {
  startISO: string
  intervalMin: number
  originLabel: string
  destinationLabel: string
}

const addMinutes = (iso: string, min: number): string =>
  new Date(new Date(iso).getTime() + min * 60_000).toISOString()

/**
 * Build the checkpoint list for a route: origin first, destination last, and an
 * evenly interpolated point every `intervalMin` minutes in between. Weather and
 * hazard are left unset — filled in later by the weather step.
 */
export function sampleCheckpoints(route: RouteResult, opts: SampleOptions): Checkpoint[] {
  const { startISO, intervalMin, originLabel, destinationLabel } = opts
  const totalMin = route.durationS / 60
  const cumM = cumulativeDistancesM(route.polyline)
  const totalKm = route.distanceM / 1000

  const offsets: number[] = [0]
  for (let t = intervalMin; t < totalMin; t += intervalMin) offsets.push(t)
  offsets.push(totalMin)

  return offsets.map((offsetMin, i): Checkpoint => {
    const isOrigin = i === 0
    const isDest = i === offsets.length - 1
    const frac = totalMin > 0 ? offsetMin / totalMin : 0
    const coord = isOrigin
      ? route.polyline[0]
      : isDest
        ? route.polyline[route.polyline.length - 1]
        : interpolateAlong(route.polyline, cumM, frac * route.distanceM)

    return {
      id: `cp-${i}`,
      kind: isOrigin ? 'origin' : isDest ? 'destination' : 'interval',
      label: isOrigin
        ? originLabel
        : isDest
          ? destinationLabel
          : `Approx. location after ${Math.round(offsetMin)} min`,
      coord,
      etaISO: addMinutes(startISO, offsetMin),
      offsetMin: Math.round(offsetMin),
      cumulativeKm: Number((frac * totalKm).toFixed(2)),
      hazard: false,
    }
  })
}
